src/sys/socket/mod.rs

```diff
@@ -774,61 +775,318 @@ impl<'a> ControlMessage<'a> {
+) -> Result<Vec<RecvMsg<'a>>>
+    where
+        I: AsRef<[IoVec<&'a mut [u8]>]> + 'a,
+{
+    let iter = data.into_iter();
+
+    let (min_size, max_size) = iter.size_hint();
+    let reserve_items = max_size.unwrap_or(min_size);
+
+    let mut output: Vec<MaybeUninit<libc::mmsghdr>> = vec![MaybeUninit::zeroed(); reserve_items];
+    let mut address: Vec<MaybeUninit<sockaddr_storage>> = vec![MaybeUninit::uninit(); reserve_items];
+
+    let results: Vec<_> = iter.enumerate().map(|(i, d)| {
+        if output.len() < i {
+            output.resize(i, MaybeUninit::zeroed());
+            address.resize(i, MaybeUninit::uninit());
+        }
+
+        let element = &mut output[i];
+
+        let msg_controllen = unsafe {
+            recv_pack_mhdr(
+                &mut (*element.as_mut_ptr()).msg_hdr,
+                d.iov.as_ref(),
+                &mut d.cmsg_buffer,
+                &mut address[i]
+            )
+        };
+
+        (msg_controllen as usize, &mut d.cmsg_buffer)
+    }).collect();
+
+    let mut initialized_data = unsafe { mem::transmute::<_, Vec<libc::mmsghdr>>(output) };
+
+    let timeout = if let Some(mut t) = timeout {
+        t.as_mut() as *mut libc::timespec
+    } else {
+        ptr::null_mut()
+    };
+
+    let ret = unsafe { libc::recvmmsg(fd, initialized_data.as_mut_ptr(), initialized_data.len() as _, flags.bits() as _, timeout) };
+
+    let r = Errno::result(ret)?;
+
+    Ok(initialized_data
+        .into_iter()
+        .zip(address.into_iter())
+        .zip(results.into_iter())
+        .map(|((mmsghdr, address), (msg_controllen, cmsg_buffer))| {
+            unsafe {
+                recv_read_mhdr(
+                    mmsghdr.msg_hdr,
+                    r as isize,
+                    msg_controllen,
+                    address,
+                    cmsg_buffer
+                )
+            }
+        })
+        .collect())
+}
+
+unsafe fn recv_read_mhdr<'a, 'b>(
+    mhdr: msghdr,
+    r: isize,
+    msg_controllen: usize,
+    address: MaybeUninit<sockaddr_storage>,
+    cmsg_buffer: &'a mut Option<&'b mut Vec<u8>>
+) -> RecvMsg<'b> {
+    let cmsghdr = {
+        if mhdr.msg_controllen > 0 {
+            // got control message(s)
+            cmsg_buffer
+                .as_mut()
+                .unwrap()
+                .set_len(mhdr.msg_controllen as usize);
+            debug_assert!(!mhdr.msg_control.is_null());
+            debug_assert!(msg_controllen >= mhdr.msg_controllen as usize);
+            CMSG_FIRSTHDR(&mhdr as *const msghdr)
+        } else {
+            ptr::null()
+        }.as_ref()
+    };
+
+    let address = sockaddr_storage_to_addr(
+        &address.assume_init(),
+         mhdr.msg_namelen as usize
+    ).ok();
+
+    RecvMsg {
+        bytes: r as usize,
+        cmsghdr,
+        address,
+        flags: MsgFlags::from_bits_truncate(mhdr.msg_flags),
+        mhdr,
+    }
+}
+
+unsafe fn recv_pack_mhdr<'a, I>(
```

Comment:
Again, I think you can avoid several problems by returning `(msghdr, sockaddr_storage)` instead of using them as return arguments.